pkg/rbac/parser.go

```diff
@@ -146,69 +147,110 @@ func (Generator) RegisterMarkers(into *markers.Registry) error {
 	return nil
 }
 
-// GenerateClusterRole generates a rbacv1.ClusterRole object
-func GenerateClusterRole(ctx *genall.GenerationContext, roleName string) (*rbacv1.ClusterRole, error) {
-	rules := make(map[ruleKey]*Rule)
+// GenerateRoles generate a slice of objs representing either a ClusterRole or a Role object
+// The order of the objs in the returned slice is stable and determined by their namespaces.
+func GenerateRoles(ctx *genall.GenerationContext, roleName string) ([]interface{}, error) {
+	rulesByNS := make(map[string][]*Rule)
 	for _, root := range ctx.Roots {
 		markerSet, err := markers.PackageMarkers(ctx.Collector, root)
 		if err != nil {
 			root.AddError(err)
 		}
 
-		// all the Rules having the same ruleKey will be merged into the first Rule
+		// group RBAC markers by namespace
 		for _, markerValue := range markerSet[RuleDefinition.Name] {
 			rule := markerValue.(Rule)
+			namespace := rule.Namespace
+			if _, ok := rulesByNS[namespace]; !ok {
+				rules := make([]*Rule, 0)
+				rulesByNS[namespace] = rules
+			}
+			rulesByNS[namespace] = append(rulesByNS[namespace], &rule)
+		}
+	}
+
+	// NormalizeRules merge Rule with the same ruleKey and sort the Rules
+	NormalizeRules := func(rules []*Rule) []rbacv1.PolicyRule {
+		ruleMap := make(map[ruleKey]*Rule)
+		// all the Rules having the same ruleKey will be merged into the first Rule
+		for _, rule := range rules {
 			key := rule.key()
-			if _, ok := rules[key]; !ok {
-				rules[key] = &rule
+			if _, ok := ruleMap[key]; !ok {
+				ruleMap[key] = rule
 				continue
 			}
-			rules[key].addVerbs(rule.Verbs)
+			ruleMap[key].addVerbs(rule.Verbs)
 		}
-	}
 
-	if len(rules) == 0 {
-		return nil, nil
-	}
+		// sort the Rules in rules according to their ruleKeys
+		keys := make([]ruleKey, 0)
```

Comment:
Change it to `make([]ruleKey, len(ruleMap))` to avoid resizing when appending.

Reply:
Done